src/filters.cpp

```diff
@@ -834,16 +834,190 @@ InternalValue Slice::Batch(const InternalValue& baseVal, RenderContext& context)
+
     return InternalValue();
 }
 
+namespace {
+
+using FormatContext = fmt::format_context;
+using FormatArgument = fmt::basic_format_arg<FormatContext>;
+
+template<typename ResultDecorator>
+struct FormatArgumentConverter : visitors::BaseVisitor<FormatArgument>
+{
+    using result_t = FormatArgument;
+
+    using BaseVisitor::operator();
+
+    FormatArgumentConverter(const RenderContext* context, const ResultDecorator& decorator)
+        : m_context(context), m_decorator(decorator)
+    {}
+
+    result_t operator()(const ListAdapter& list) const
+    {
+        return make_result(Apply<PrettyPrinter>(list, m_context));
+    }
+
+    result_t operator()(const MapAdapter& map) const
+    {
+        return make_result(Apply<PrettyPrinter>(map, m_context));
+    }
+
+    result_t operator()(const std::string& str) const
+    {
+        return make_result(str);
+    }
+
+    result_t operator()(const nonstd::string_view& str) const
+    {
+        return make_result(std::string(str.data(), str.size()));
+    }
+
+    result_t operator()(const std::wstring& str) const
+    {
+        return make_result(ConvertString<std::string>(str));
+    }
+
+    result_t operator()(const nonstd::wstring_view& str) const
+    {
+        return make_result(ConvertString<std::string>(str));
+    }
+
+    result_t operator()(double val) const
+    {
+        return make_result(val);
+    }
+
+    result_t operator()(int64_t val) const
+    {
+        return make_result(val);
+    }
+
+    result_t operator()(bool val) const
+    {
+        return make_result(val ? "true"s : "false"s);
+    }
+
+    result_t operator()(EmptyValue) const
+    {
+        return make_result("none"s);
+    }
+
+    result_t operator()(const Callable&) const
+	{
+		return make_result("<callable>"s);
+	}
+
+    template<typename T>
+    result_t make_result(const T& t) const
+    {
+        return fmt::internal::make_arg<FormatContext>(m_decorator(t));
+    }
+
+
+    const RenderContext* m_context;
+    const ResultDecorator& m_decorator;
+};
+
+template<typename T>
+using NamedArgument = fmt::internal::named_arg<T, char>;
+
+using ValueHandle = nonstd::variant<
+    bool,
+    std::string,
+    int64_t,
+    double,
+    NamedArgument<bool>,
+    NamedArgument<std::string>,
+    NamedArgument<int64_t>,
+    NamedArgument<double>
+>;
+using ValuesBuffer = std::vector<ValueHandle>;
+
+struct CachingIdentity
+{
+public:
+    CachingIdentity(ValuesBuffer& values) : m_values(values)
+    {}
+
+    template<typename T>
+    const auto& operator()(const T& t) const
+    {
+        m_values.push_back(t);
+        return m_values.back().get<T>();
+    }
+private:
+    ValuesBuffer& m_values;
+};
+
+class NamedArgumentCreator
+{
+public:
+    NamedArgumentCreator(const std::string& name, ValuesBuffer& valuesBuffer)
+        : m_name(name), m_valuesBuffer(valuesBuffer)
+    {}
+
+    template<typename T>
+    const auto& operator()(const T& t) const
```

Comment:
I suppose, the problem with `use-after-free` error is here. When you push into vector and reallocation occurred (which is possible even with reserve) all references to the previous items are invalidated.

Reply:
The funny thing is that, I cannot get sanitizer problems on my Ubuntu vm. VS2019 won't find any problems either. I believe that push_backs shouldn't do reallocation: https://en.cppreference.com/w/cpp/container/vector/push_back, https://stackoverflow.com/a/10735186. I will look into it a little more.